diff --git a/kexec/kexec-syscall.h b/kexec/kexec-syscall.h
index f4df45c..4d74c51 100644
--- a/kexec/kexec-syscall.h
+++ b/kexec/kexec-syscall.h
@@ -75,6 +75,14 @@ static inline long kexec_reboot(void)
 	return (long) syscall(__NR_reboot, LINUX_REBOOT_MAGIC1, LINUX_REBOOT_MAGIC2, LINUX_REBOOT_CMD_KEXEC, 0);
 }
 
+/* MKLINUX */
+#define __NR_multikernel_boot 312
+#define LINUX_REBOOT_CMD_MKBSP 0xDEAD5E55
+static inline long kexec_boot(int cpu, unsigned long kernel_start_address)
+{
+	return (long) syscall(__NR_multikernel_boot, cpu, kernel_start_address);
+}
+
 
 #define KEXEC_ON_CRASH		0x00000001
 #define KEXEC_PRESERVE_CONTEXT	0x00000002